app/modules/sntppkt.c

```diff
@@ -0,0 +1,315 @@
+
+  if (memcmp((const char *)cookie, (const char *)&pktb.origin, sizeof (*cookie))) {
+    /* bad cookie; return nil */
+    return 0;
+  }
+
+  /* KOD? */
+  if (pktb.LI == 3) {
+    lua_pushlstring(L, (const char *)&pktb.refid, 4);
+    return 1;
+  }
+
+  ntpr = lua_newuserdata(L, sizeof(ntp_response_t));
+  luaL_getmetatable(L, "sntppkt.resp");
+  lua_setmetatable(L, -2);
+
+  ntpr->LI              = pktb.LI;
+  ntpr->stratum         = pktb.stratum;
+  ntpr->root_delay      = ntohl(pktb.root_delay);
+  ntpr->root_dispersion = ntohl(pktb.root_dispersion);
+
+  /* Heavy time lifting time */
+
+  pktb.origin.sec  = ntohl(pktb.origin.sec);
+  pktb.origin.frac = ntohl(pktb.origin.frac);
+  pktb.recv.sec    = ntohl(pktb.recv.sec);
+  pktb.recv.frac   = ntohl(pktb.recv.frac);
+  pktb.xmit.sec    = ntohl(pktb.xmit.sec);
+  pktb.xmit.frac   = ntohl(pktb.xmit.frac);
+
+  ntpr->txsec = pktb.xmit.sec - NTP_TO_UNIX_EPOCH;
+
+  uint64_t ntp_recv   = (((uint64_t) pktb.recv.sec               ) << 32)
+                        + pktb.recv.frac;
+  uint64_t ntp_origin = (((uint64_t) pktb.origin.sec             ) << 32)
+                        + pktb.origin.frac;
+  uint64_t ntp_xmit   = (((uint64_t) pktb.xmit.sec               ) << 32)
+                        + pktb.xmit.frac;
+  uint64_t ntp_dest   = (((uint64_t) now_sec + NTP_TO_UNIX_EPOCH ) << 32)
+                        + sntppkt_us_to_frac(now_usec);
+
+  ntpr->delta = ((int64_t) ntp_recv - ntp_origin) / 2
+              + ((int64_t) ntp_xmit - ntp_dest  ) / 2;
+
+  ntpr->delay_frac = ((int64_t)ntp_dest - ntp_origin - ntp_xmit + ntp_recv) >> 16;
+
+  ntpr->cached_delay    = ntpr->root_delay * 2 + ntpr->delay_frac;
+
+  return 1;
+}
+
+/*
+ * Left-biased selector of a "preferred" NTP response.  Note that preference
+ * is rather subjective!
+ *
+ * Lua does not make it straightforward to return an existing userdata
+ * object, so instead we merely return a boolean indicating whether the
+ * second argument is superior to the first.
+ */
+
+static int
+sntppkt_pick_resp(lua_State *L) {
+
+  ntp_response_t *a = luaL_checkudata(L, 1, "sntppkt.resp");
+  ntp_response_t *b = luaL_checkudata(L, 2, "sntppkt.resp");
+  int biased = 0;
+
+  biased = lua_toboolean(L, 3);
+
+  /*
+   * If we're "biased", prefer the second structure if the delay less than
+   * 3/4ths of the delay in the first.  An unbiased comparison just uses
+   * the raw delay values.
+   */
+  if (biased) {
+    lua_pushboolean(L, a->cached_delay * 3 > b->cached_delay * 4);
+  } else {
+    lua_pushboolean(L, a->cached_delay     > b->cached_delay    );
+  }
+  return 1;
+}
+
+/*
+ * Inflate a NTP response into a Lua table
+ */
+static int
+sntppkt_read_resp(lua_State *L) {
+  ntp_response_t *r = luaL_checkudata(L, 1, "sntppkt.resp");
+
+  lua_createtable(L, 0, 6);
+
+  /* For large corrections, don't bother exposing fine values */
+  int d40 = r->delta >> 40;
+  if (d40 != 0 && d40 != -1) {
+    lua_pushnumber(L, r->delta >> 32);
```

Comment:
You use this pattern a lot.  Why not just have function which adds an integer field to the array at ToS?
```C
static void numuber_tofield(lua_State *L, const char * field_name, lua_Number value) {
  lua_pushnumber(L, value);
  lua_setfield(L, -2, field_name);
}
```
This will generate a little less code.